### PR TITLE
[HOPSWORKS-2827] Add serving module

### PR DESCRIPTION
**Pull request by javierdlrm**

### CODE REVIEW

File: python/hsml/sklearn/predictor_config.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+from hsml.predictor_config import PredictorConfig
+
+
+class PredictorConfig(PredictorConfig):
+    """Configuration for a predictor running a sklearn model."""
+
+    def __init__(
+        self,
+        serving_tool="DEFAULT",
```
Comment (robzor92) [inline python/hsml/sklearn/predictor_config.py:25]: Could use the class PREDICTOR constants from constants.py instead for these files.

File: python/hsml/transformer_config.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#
+#   Copyright 2021 Logical Clocks AB
```
Comment (robzor92) [inline python/hsml/transformer_config.py:2]: 2022

File: python/hsml/util.py [hunk 3]
```diff
@@ -189,3 +196,27 @@ def validate_metric_value(opt_val):
     raise TypeError(
         "Metric value is of type {}, expecting a number".format(type(opt_val))
     )
+
+
+def get_predictor_config_for_model(model):
+    if model.framework is None:
```
Comment (robzor92) [inline python/hsml/util.py:202]: I guess this is a model object, in that case you check the model is of type BaseClass etc.

**Review by robzor92 — CHANGES_REQUESTED** [review]

File: python/hsml/client/hopsworks/base.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import os
+from abc import abstractmethod
+
+from hsml.client import base, auth
+
+
+class Client(base.Client):
+    TOKEN_FILE = "token.jwt"
+    REST_ENDPOINT = "REST_ENDPOINT"
+    DEFAULT_DATABRICKS_ROOT_VIRTUALENV_ENV = "DEFAULT_DATABRICKS_ROOT_VIRTUALENV_ENV"
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/client/hopsworks/base.py:26]: Done

File: python/hsml/client/istio/base.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    def _get_host_port_pair(self):
+        """
+        Removes "http or https" from the rest endpoint and returns a list
+        [endpoint, port], where endpoint is on the format /path.. without http://
+
+        :return: a list [endpoint, port]
+        :rtype: list
+        """
+        endpoint = self._base_url
+        if "http" in endpoint:
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/client/istio/base.py:76]: I've changed it here. But it comes from [hsfs](https://github.com/logicalclocks/feature-store-api/blob/4a2c4b2f4291623c62280bdd8a757c02754ec5d2/python/hsfs/client/base.py#L75).

File: python/hsml/client/istio/internal.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        self._auth = auth.ApiKeyAuth(self._get_serving_api_key())
+        self._verify = self._get_verify(hostname_verification, trust_store_path)
+        self._session = requests.session()
+
+        self._connected = True
+
+    def _get_istio_endpoint(self):
+        """Get the istio endpoint for making requests to the ingress gateway."""
+        if self.ISTIO_ENDPOINT not in os.environ:
+            raise exceptions.ExternalClientError("Istio endpoint not found")
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/client/istio/internal.py:72]: Fixed

File: python/hsml/client/istio/internal.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            "__"
+        )  # project users have username project__user
+        project = hops_user_split[0]
+        return project
+
+    def _project_user(self):
+        try:
+            hops_user = os.environ[self.HADOOP_USER_NAME]
+        except KeyError:
+            hops_user = os.environ[self.HDFS_USER]
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/client/istio/internal.py:92]: It makes sense

File: python/hsml/component_config.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    def __init__(
+        self,
+        script_file: Optional[str] = None,
+        resources_config: Optional[Union[ResourcesConfig, dict]] = None,
+        inference_logger: Optional[Union[InferenceLoggerConfig, dict]] = None,
+        inference_batcher: Optional[Union[InferenceBatcherConfig, dict]] = None,
+    ):
+        # check for dict params
+        resources_config = util.get_obj_from_json(
+            ResourcesConfig, resources_config, self.REQUESTED_INSTANCES_KEY
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/component_config.py:43]: I think we don't need to do that in this case. So far, in the backend resources config contains cores, memory and gpus. In HSML, I moved number of instances to resourcesConfig class for the sake of clarity, trying to organize the different types of configuration we have. But when we serialize/deserialize, number of instances is treated as a field outside resources_config (with names `requested_instances ` and `requested_transformer_instances`). This key is known by the predictorConfig and transformerConfig, not the resourcesConfig.
However, a better approach can be found in a following PR.

File: python/hsml/inference_batcher_config.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import json
+import humps
+from typing import Optional
+
+from hsml import util
+
+from hsml.constants import INFERENCE_BATCHER
+
+
+class InferenceBatcherConfig:
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/inference_batcher_config.py:25]: Yes, there are more configuration parameters that are supported in KServe (and we will add support for) such as max_batch_size.

File: python/hsml/model.py [hunk 10]
```diff
@@ -308,6 +329,13 @@ def version_path(self):
         """path of the model including version folder. Resolves to /Projects/{project_name}/Models/{name}/{version}"""
         return "{}/{}".format(self.model_path, str(self.version))
 
+    @property
+    def absolute_path(self):
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/model.py:333]: It's not needed anymore after adding `model_path`. I have removed it.

File: python/hsml/sklearn/predictor_config.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        script_file=None,
+        resources_config=None,
+        inference_logger=None,
+        inference_batcher=None,
+    ):
+        super().__init__(
+            model_server="PYTHON",
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/sklearn/predictor_config.py:32]: Done

File: python/hsml/sklearn/predictor_config.py (hunk 1 above)
Comment (javierdlrm) [reply python/hsml/sklearn/predictor_config.py:25]: Done

File: python/hsml/transformer_config.py (hunk 2 above)
Comment (javierdlrm) [reply python/hsml/transformer_config.py:2]: Fixed

File: python/hsml/util.py (hunk 3 above)
Comment (javierdlrm) [reply python/hsml/util.py:202]: Done

File: python/hsml/connection.py [hunk 12]
```diff
@@ -137,6 +138,19 @@ def get_model_registry(self, project: str = None):
         """
         return self._model_registry_api.get(project)
 
+    @connected
+    def get_model_serving(self, project: str = None):
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/connection.py:142]: That's correct. I've refactored it.

File: python/hsml/core/model_serving_api.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+from hsml.model_serving import ModelServing
+from hsml.core import dataset_api
+from hsml.client.exceptions import ModelRegistryException
+
+
+class ModelServingApi:
+    def __init__(self):
+        self._dataset_api = dataset_api.DatasetApi()
+
+    def get(self, project=None):
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsml/core/model_serving_api.py:27]: Fixed

File: python/hsml/model_serving.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+class ModelServing:
+    DEFAULT_VERSION = 1
+
+    def __init__(self, project_name: str, project_id: int):
+        self._project_name = project_name
+        self._project_id = project_id
+
+        self._serving_api = serving_api.ServingApi()
+
+    def get_deployment(self, id: int):
```
Comment (robzor92) [inline python/hsml/model_serving.py:36]: Maybe nicer UX to have a get_deployment_by_id and have the get_deployment instead take as a parameter the name. I think most users will want to pass the name directly.

**Review by robzor92 — CHANGES_REQUESTED** [review]

File: python/hsml/util.py [hunk 15]
```diff
@@ -189,3 +198,34 @@ def validate_metric_value(opt_val):
...
+    if not isinstance(model, BaseModel):
+        raise ValueError(
+            "model is of type {}, but an instance of {} class is expected".format(
+                type(model), BaseModel
+            )
+        )
+
+    if model.framework is None:
+        return BasePredictorConfig(model_server=PREDICTOR.MODEL_SERVER_PYTHON)
+    if model.framework == "TENSORFLOW":
```
Comment (robzor92) [inline python/hsml/util.py:213]: Also here you can compare the class directly. `type(model) == TFModel` for example.

**Review by robzor92 — CHANGES_REQUESTED** [review]

File: python/hsml/model_serving.py (hunk 14 above)
Comment (javierdlrm) [reply python/hsml/model_serving.py:36]: Yes, good point.

File: python/hsml/util.py (hunk 15 above)
Comment (javierdlrm) [reply python/hsml/util.py:213]: Changed

File: python/hsml/util.py [hunk 16]
```diff
@@ -189,3 +198,34 @@ def validate_metric_value(opt_val):
...
+
+def get_predictor_config_for_model(model: BaseModel):
+    if not isinstance(model, BaseModel):
+        raise ValueError(
+            "model is of type {}, but an instance of {} class is expected".format(
+                type(model), BaseModel
+            )
+        )
+
+    if type(model) == TFModel:
```
Comment (robzor92) [inline python/hsml/util.py:211]: These statements could be an if elif elif etc

**Review by robzor92 — APPROVED** [review]

**Pull request merged**